File: SourceAFIS/Extraction/Templates/Template.cs
using System;
using System.Collections.Generic;
using System.Text;
using SourceAFIS.General;
using SourceAFIS.Dummy;

namespace SourceAFIS.Extraction.Templates
{
    [Serializable]
    public sealed class Template : ICloneable
    {
        public enum MinutiaType : byte
        {
            Ending = 0,
            Bifurcation = 1
        }

        [Serializable]
        public struct Minutia
        {
            public readonly PointS Position;
            public readonly byte Direction;
            public readonly MinutiaType Type;

            public Minutia(TemplateBuilder.Minutia builderMinutia)
            {
                Position = new PointS(builderMinutia.Position);
                Direction = builderMinutia.Direction;
                if (builderMinutia.Type == TemplateBuilder.MinutiaType.Ending)
                    Type = MinutiaType.Ending;
                else
                    Type = MinutiaType.Bifurcation;
            }

            public TemplateBuilder.Minutia ToBuilderMinutia()
            {
                TemplateBuilder.Minutia builderMinutia = new TemplateBuilder.Minutia();
                builderMinutia.Position = Position;
                builderMinutia.Direction = Direction;
                if (Type == MinutiaType.Ending)
                    builderMinutia.Type = TemplateBuilder.MinutiaType.Ending;
                else
                    builderMinutia.Type = TemplateBuilder.MinutiaType.Bifurcation;
                return builderMinutia;
            }
        }

        public readonly Minutia[] Minutiae;

        public Template(TemplateBuilder builder)
        {
            Minutiae = new Minutia[builder.Minutiae.Count];
            for (int i = 0; i < builder.Minutiae.Count; ++i)
                Minutiae[i] = new Minutia(builder.Minutiae[i]);
        }

        public TemplateBuilder ToTemplateBuilder()
        {
            TemplateBuilder builder = new TemplateBuilder();
            foreach (Minutia minutia in Minutiae)
                builder.Minutiae.Add(minutia.ToBuilderMinutia());
            return builder;
        }

        public Template Clone()
        {
            return new Template(ToTemplateBuilder());
        }

        object ICloneable.Clone() { return Clone(); }
    }
}

File: java/sourceafis/src/sourceafis/extraction/templates/Template.java
package sourceafis.extraction.templates;

import java.io.Serializable;

  

@SuppressWarnings("serial")
public class Template implements Cloneable, Serializable {
 
	public int OriginalDpi;
	public int OriginalWidth;
	public int OriginalHeight;
	public int StandardDpiWidth;
	public int StandardDpiHeight;
	public Minutia[] Minutiae;
   /*
	public class Minutia {
		public PointS Position;
		public byte Direction;
		public MinutiaType Type;

		/*public Minutia(TemplateBuilder.Minutia builderMinutia) {
			Position = new PointS(builderMinutia.Position);
			Direction = builderMinutia.Direction;
			if (builderMinutia.Type == MinutiaType.Ending)
				Type = MinutiaType.Ending;
			else
				Type = MinutiaType.Bifurcation;
		}

		public TemplateBuilder.Minutia ToBuilderMinutia() {
			 TemplateBuilder.Minutia builderMinutia = new  TemplateBuilder.Minutia();// new
																							// TemplateBuilder.Minutia();
			builderMinutia.Position = Position.ToPoint();
			builderMinutia.Direction = Direction;
			if (Type == MinutiaType.Ending)
				builderMinutia.Type = MinutiaType.Ending;
			else
				builderMinutia.Type = MinutiaType.Bifurcation;
			return builderMinutia;
		} 
	}*/
	
	public Template(TemplateBuilder builder) {
		OriginalDpi = builder.OriginalDpi;
		OriginalWidth = builder.OriginalWidth;
		OriginalHeight = builder.OriginalHeight;
		StandardDpiWidth = builder.getStandardDpiWidth();
		StandardDpiHeight = builder.getStandardDpiHeight();

		Minutiae = new Minutia[builder.Minutiae.size()];
		for (int i = 0; i < builder.Minutiae.size(); ++i)
			Minutiae[i] = builder.Minutiae.get(i);//new Minutia(builder.Minutiae.get(i));
	}

	public TemplateBuilder ToTemplateBuilder() {
		TemplateBuilder builder = new TemplateBuilder();

		builder.OriginalDpi = OriginalDpi;
		builder.OriginalWidth = OriginalWidth;
		builder.OriginalHeight = OriginalHeight;
		for (Minutia minutia : Minutiae) {
			//builder.Minutiae.add(minutia.ToBuilderMinutia());
			builder.Minutiae.add(minutia);
		}
		return builder;
	}

	public Template clone() {
		return new Template(ToTemplateBuilder());
	}

 
}
